### ISSUE TITLE
learning rate warmup example in the document does not change the learning rate of the first epoch

### ISSUE THREAD

**Issue by zhong-yy** [Lightning-AI/pytorch-lightning#17972]
### Bug description

I used the example in the document to perform learning rate warmup.
https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#lightningmodule-api

But it seems that the warmup only starts from the second epoch
![output](https://github.com/Lightning-AI/lightning/assets/38816011/3dd6c43d-c3e4-4a76-b971-47e924e09efc)

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
# Learning rate warm-up
def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_closure):
    # update params
    optimizer.step(closure=optimizer_closure)

    # manually warm up lr without a scheduler
    if self.trainer.global_step < 500:
        lr_scale = min(1.0, float(self.trainer.global_step + 1) / 500.0)
        for pg in optimizer.param_groups:
            pg["lr"] = lr_scale * self.learning_rate
```

My optimizer is like this:
```python
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        if self.lr_scheduler:
            learning_rate_scheduler=torch.optim.lr_scheduler.__getattribute__(self.lr_scheduler)(optimizer=optimizer,**self.lr_scheduler_args)
            return {"optimizer":optimizer,
                    "lr_scheduler":{
                        # REQUIRED: The scheduler instance
                        "scheduler": learning_rate_scheduler,
                        # The unit of the scheduler's step size, could also be 'step'.
                        # 'epoch' updates the scheduler on epoch end whereas 'step'
                        # updates it after a optimizer update.
                        "interval": "epoch",
                        # How many epochs/steps should pass between calls to
                        # `scheduler.step()`. 1 corresponds to updating the learning
                        # rate after every epoch/step.
                        "frequency": 1,
                        # Metric to to monitor for schedulers like `ReduceLROnPlateau`
                        "monitor": "val_loss",
                        # If set to `True`, will enforce that the value specified 'monitor'
                        # is available when the scheduler is updated, thus stopping
                        # training if not found. If set to `False`, it will only produce a warning
                        "strict": False,
                        # If using the `LearningRateMonitor` callback to monitor the
                        # learning rate progress, this keyword can be used to specify
                        # a custom logged name
                        "name": None,
                    }         
                }
        else:
            return optimizer
```
```


### Error messages and logs

```
# Error messages and logs here please
```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>


### More info

_No response_

cc @borda

**Comment by function2-llx** [comment]
I've got the same concern. I also observe this when "interval" is "step" and the learning rate for the first step is not initialized properly. This is probably due to lr scheduling is always performed after the optimization step or after an epoch is done.
https://github.com/Lightning-AI/lightning/blob/234a78a957e4adf288e7a30e9c9092fc6f1febd5/src/lightning/pytorch/loops/training_epoch_loop.py#L218-L231

**Comment by awaelchli** [comment]
I don't like this example we have in the docs. LR warmup should be done via a LR scheduler and be configured in the `configure_optimizers()` hook. I suggest we remove this example as a use case for the `optimizer_step` hook.

My two cents on the second point: I disgree that stepping the scheduler after the optimizer step is misbehavior. There are plenty of examples out there that show the scheduler being called after the optimizer step, and that is considered the standard practice. Besides it being a breaking change, doing it the other way around would only raise questions by users who look at our implementation. Furthermore, if you prefer to have the learning rate at a particular value before the very first optimizer step, then just set this as the initial learning rate in the optimizer configuration. If that is still not satisfactory, then manual optimization mode is the way to go :)

**Comment by zhong-yy** [comment]
@awaelchli Thanks for your reply. Could you give an example of configuring LR warmup?

**Comment by Quantum1921** [comment]
No ..

**Comment by function2-llx** [comment]
@zhong-yy Have you tried LR scheduler from timm? It will initialize the LR to some value when using warmup, e.g.:

https://github.com/huggingface/pytorch-image-models/blob/ef72c3cd470dd67836eebf95ec567199c890a6a2/timm/scheduler/cosine_lr.py#L74-L76
